### ISSUE TITLE
Problems trying to visualize Network Training with the Deeplearning4j Training UI

### ISSUE THREAD

**Issue by vincenzocaselli** [deeplearning4j/deeplearning4j#8693]
I tried to follow the guide here
https://deeplearning4j.org/docs/latest/deeplearning4j-nn-visualization
but during training there is apparently nothing published at
http://localhost:9000/train

I tried to apply the guide to an existing and working project with the following initial pom.xml dependencies:
```
		<dependency>
			<groupId>org.nd4j</groupId>
			<artifactId>nd4j-native-platform</artifactId>
			<version>1.0.0-beta6</version>
		</dependency>
		<dependency>
			<groupId>org.deeplearning4j</groupId>
			<artifactId>deeplearning4j-core</artifactId>
			<version>1.0.0-beta6</version>
		</dependency>

```
Following the deeplearning4j-nn-visualization page I tried to add this dependency
```
    <dependency>
        <groupId>org.deeplearning4j</groupId>
        <artifactId>deeplearning4j-ui_2.10</artifactId>
        <version>1.0.0-beta4</version>
    </dependency>

```
But simply doing, when executing line
`UIServer uiServer = UIServer.getInstance();`
I get the following error
```
Exception in thread "main" java.lang.IllegalArgumentException: Don't know how to bind argument of type class net.jodah.typetools.TypeResolver$Unknown
	at play.routing.RoutingDsl.pathBindableFor(RoutingDsl.java:233)
	at play.routing.RoutingDsl.with(RoutingDsl.java:190)
	at play.routing.RoutingDsl.access$000(RoutingDsl.java:66)
	at play.routing.RoutingDsl$PathPatternMatcher.build(RoutingDsl.java:362)
	at play.routing.RoutingDsl$PathPatternMatcher.routeTo(RoutingDsl.java:298)
	at org.deeplearning4j.ui.play.PlayUIServer.runMain(PlayUIServer.java:179)
	at org.deeplearning4j.ui.api.UIServer.getInstance(UIServer.java:67)
	at org.deeplearning4j.ui.api.UIServer.getInstance(UIServer.java:45)

```
Then I tried to add
```
		<dependency>
			<groupId>org.deeplearning4j</groupId>
			<artifactId>deeplearning4j-ui</artifactId>
			<version>1.0.0-beta6</version>
		</dependency>

```
with this, there are no errors, but if during the training I open browser (as indicated in the guide) at
http://localhost:9000/train
the page is unreachable.

There is some other dependency to be added?

I am using the following versions:

* Deeplearning4j version:  1.0.0-beta6
* Platform information (OS, etc): Linux Arch 64bits
* CUDA version: not used
* NVIDIA driver version: not used

Thank you very much
Best regards
Vincenzo

**Comment by AlexDBlack** [comment]
First - don't use mixed versions, everything (all ND4J/DL4J/DataVec etc dependencies) should always be the same version. You probably got a warning about that.
(But yes, that page is outdated listing 1.0.0-beta4 deeplearning4j-ui_2.10 instead of 1.0.0.-beta6 deeplearning4j-ui, cc @treo )

As for not being able to see anything: I think that's probably code related.
The UI server doesn't initialize (i.e., no web server) until you do this:
```
UIServer uiServer = UIServer.getInstance();
```
and you won't have any data until you do this: (you'll see the empty graphs/layout though)
```
uiServer.attach(statsStorage);
```

I'm pretty sure being unable to access the server is due to no initialization of the UI server. If you still can't get it working, post your code too and we can re-open this issue and help you more.

**Comment by vincenzocaselli** [comment]
Hi Alex,
I am now using 1.0.0-beta6 for all dependencies.
I followed exactly the code in the guide, that is:
```
    //Initialize the user interface backend
    UIServer uiServer = UIServer.getInstance();

    //Configure where the network information (gradients, score vs. time etc) is to be stored. Here: store in memory.
    StatsStorage statsStorage = new InMemoryStatsStorage();         //Alternative: new FileStatsStorage(File), for saving and loading later

    //Attach the StatsStorage instance to the UI: this allows the contents of the StatsStorage to be visualized
    uiServer.attach(statsStorage);

    //Then add the StatsListener to collect this information from the network, as it trains
    net.setListeners(new StatsListener(statsStorage));

```
but things doesn't change.
I also was able to reproduce the behavior cloning repo
https://github.com/eclipse/deeplearning4j-examples.git
and executing class
**org.deeplearning4j.examples.userInterface.util.UIExample.java** 
The execution hangs at line
`net.fit(trainData);`
where the console shows
```
Feb 10, 2020 12:34:02 AM org.nd4j.util.OneTimeLogger info
INFO: Starting MultiLayerNetwork with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
Feb 10, 2020 12:34:03 AM org.deeplearning4j.ui.VertxUIServer attach
INFO: StatsStorage instance attached to UI: FileStatsStorage(/tmp/ui-stats.dl4j)

```
and address 
http://localhost:9000/train
is unreachable.
Is the address right? What could cause this block?
Just a note: if I comment those UIServer-related lines of code all is working right (no freeze)

**Comment by AlexDBlack** [comment]
> http://localhost:9000/train is unreachable. Is the address right?

Yes, that's the correct address.

That's unexpected, there are no known issues with freezing during training, with or without the UI.
Furthermore even if there was an issue during training, it should not make any difference what-so-ever to the UI web server - they are relatively independent systems.

Can you confirm that the examples were a fresh clone, completely unmodified?
And is this being run in an IDE or via maven command line?

We'll also need more info about your system. OS, basic hardware, etc.

Finally, can you run it until it freezes, and then give us a thread dump?
If you're using Intellij, you can use "Dump Threads" in the run tool window - https://www.jetbrains.com/help/idea/run-tool-window.html
If you're on command line, you can use jps and jstack - https://helpx.adobe.com/au/experience-manager/kb/TakeThreadDump.html

**Comment by saudet** [comment]
> Platform information (OS, etc): Linux Arch 64bits

I'm pretty sure those builds are not up and running yet @sshepel?

Not seeing anything here, for example:
https://jenkins.konduit.ai/blue/organizations/jenkins/eclipse%2Fdeeplearning4j/detail/master/77/pipeline

**Comment by vincenzocaselli** [comment]
Hi Alex,
git is up-to-date (commit 02c1489... as of 2020-02-05), no unstaged/staged changes.

Class **org.deeplearning4j.examples.userInterface.UIExample.java** is launched from **Eclipse** IDE with JDK **java-13-openjdk**.

I am using **Linux Arch 64-bits** (Manjaro) on Dell Precision 5530.
The launch hangs with these messages in console:

```
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/vincenzo/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.25/8dacf9514f0c707cbbcdd6fd699e8940d42fb54e/slf4j-simple-1.7.25.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/vincenzo/.m2/repository/ch/qos/logback/logback-classic/1.1.7/logback-classic-1.1.7.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.SimpleLoggerFactory]
[main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for linear algebra: 6
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - *********************************** CPU Feature Check Warning ***********************************
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - Warning: Initializing ND4J with Generic x86 binary on a CPU with AVX/AVX2 support
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - Using ND4J with AVX/AVX2 will improve performance. See deeplearning4j.org/cpu for more details
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - Or set environment variable ND4J_IGNORE_AVX=true to suppress this warning
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - *************************************************************************************************
[main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for OpenMP BLAS: 6
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Linux]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [12]; Memory: [3.8GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
[main] INFO org.deeplearning4j.nn.multilayer.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
[main] INFO org.deeplearning4j.ui.VertxUIServer - StatsStorage instance attached to UI: FileStatsStorage(/tmp/ui-stats.dl4j)
```
At this point address 
http://localhost:9000/train
is unreachable.

This is the thread dump:

```
[vincenzo@ARCH-P5530 ~]$ /usr/lib/jvm/java-13-openjdk/bin/jstack 5419
2020-02-12 00:00:06
Full thread dump OpenJDK 64-Bit Server VM (13.0.2+8 mixed mode):

Threads class SMR info:
_java_thread_list=0x00007fddb4001f50, length=20, elements={
0x00007fde04027000, 0x00007fde04277800, 0x00007fde0427b000, 0x00007fde0429c800,
0x00007fde04291800, 0x00007fde04293800, 0x00007fde04295800, 0x00007fde04314800,
0x00007fde0439f800, 0x00007fde043a3000, 0x00007fdd9c001800, 0x00007fde056d3800,
0x00007fde05e9c800, 0x00007fde063dd800, 0x00007fde06895000, 0x00007fdd1009b000,
0x00007fde06b68800, 0x00007fdd08005000, 0x00007fdd08006800, 0x00007fddb4001000
}

"main" #1 prio=5 os_prio=0 cpu=72595,89ms elapsed=80,20s tid=0x00007fde04027000 nid=0x1533 runnable  [0x00007fde084fb000]
   java.lang.Thread.State: RUNNABLE
	at org.nd4j.nativeblas.Nd4jCpu.execSummaryStatsScalar(Native Method)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:404)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:246)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.std(BaseNDArray.java:3967)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.std(BaseNDArray.java:3961)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.std(BaseNDArray.java:3956)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.stdNumber(BaseNDArray.java:1213)
	at org.deeplearning4j.ui.stats.BaseStatsListener.calculateSummaryStats(BaseStatsListener.java:744)
	at org.deeplearning4j.ui.stats.BaseStatsListener.onForwardPass(BaseStatsListener.java:272)
	at org.deeplearning4j.ui.stats.BaseStatsListener.onForwardPass(BaseStatsListener.java:256)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2754)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2708)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:170)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fitHelper(MultiLayerNetwork.java:1713)
	- locked <0x000000070c48c3b8> (a org.deeplearning4j.nn.multilayer.MultiLayerNetwork)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1634)
	at org.deeplearning4j.examples.userInterface.UIExample.main(UIExample.java:57)

"Reference Handler" #2 daemon prio=10 os_prio=0 cpu=145,04ms elapsed=80,18s tid=0x00007fde04277800 nid=0x153b waiting on condition  [0x00007fddd3275000]
   java.lang.Thread.State: RUNNABLE
	at java.lang.ref.Reference.waitForReferencePendingList(java.base@13.0.2/Native Method)
	at java.lang.ref.Reference.processPendingReferences(java.base@13.0.2/Reference.java:241)
	at java.lang.ref.Reference$ReferenceHandler.run(java.base@13.0.2/Reference.java:213)

"Finalizer" #3 daemon prio=8 os_prio=0 cpu=1,65ms elapsed=80,18s tid=0x00007fde0427b000 nid=0x153c in Object.wait()  [0x00007fddd3174000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@13.0.2/Native Method)
	- waiting on <0x000000070bbfb4d0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:155)
	- locked <0x000000070bbfb4d0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:176)
	at java.lang.ref.Finalizer$FinalizerThread.run(java.base@13.0.2/Finalizer.java:170)

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 cpu=0,33ms elapsed=80,17s tid=0x00007fde0429c800 nid=0x153d runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=0 cpu=16431,87ms elapsed=80,17s tid=0x00007fde04291800 nid=0x153e waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   No compile task

"C1 CompilerThread0" #8 daemon prio=9 os_prio=0 cpu=2603,27ms elapsed=80,17s tid=0x00007fde04293800 nid=0x153f waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   No compile task

"Sweeper thread" #9 daemon prio=9 os_prio=0 cpu=72,10ms elapsed=80,17s tid=0x00007fde04295800 nid=0x1540 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Common-Cleaner" #10 daemon prio=8 os_prio=0 cpu=1,88ms elapsed=80,09s tid=0x00007fde04314800 nid=0x1542 in Object.wait()  [0x00007fddd2988000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(java.base@13.0.2/Native Method)
	- waiting on <0x000000070af00178> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:155)
	- locked <0x000000070af00178> (a java.lang.ref.ReferenceQueue$Lock)
	at jdk.internal.ref.CleanerImpl.run(java.base@13.0.2/CleanerImpl.java:148)
	at java.lang.Thread.run(java.base@13.0.2/Thread.java:830)
	at jdk.internal.misc.InnocuousThread.run(java.base@13.0.2/InnocuousThread.java:134)

"JDWP Transport Listener: dt_socket" #11 daemon prio=10 os_prio=0 cpu=13,25ms elapsed=80,06s tid=0x00007fde0439f800 nid=0x1545 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"JDWP Event Helper Thread" #12 daemon prio=10 os_prio=0 cpu=1,36ms elapsed=80,06s tid=0x00007fde043a3000 nid=0x1547 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"JDWP Command Reader" #13 daemon prio=10 os_prio=0 cpu=7,47ms elapsed=80,06s tid=0x00007fdd9c001800 nid=0x1549 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Service Thread" #14 daemon prio=9 os_prio=0 cpu=0,38ms elapsed=79,60s tid=0x00007fde056d3800 nid=0x1552 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"JavaCPP Deallocator" #16 daemon prio=10 os_prio=0 cpu=216,83ms elapsed=78,80s tid=0x00007fde05e9c800 nid=0x155f in Object.wait()  [0x00007fdd5efc2000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@13.0.2/Native Method)
	- waiting on <0x000000070c48c698> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:155)
	- locked <0x000000070c48c698> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:176)
	at org.bytedeco.javacpp.Pointer$DeallocatorThread.run(Pointer.java:375)

"vertx-blocked-thread-checker" #17 daemon prio=5 os_prio=0 cpu=6,52ms elapsed=77,21s tid=0x00007fde063dd800 nid=0x157b in Object.wait()  [0x00007fdd343e3000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(java.base@13.0.2/Native Method)
	- waiting on <0x000000070ea3cbb8> (a java.util.TaskQueue)
	at java.util.TimerThread.mainLoop(java.base@13.0.2/Timer.java:553)
	- locked <0x000000070ea3cbb8> (a java.util.TaskQueue)
	at java.util.TimerThread.run(java.base@13.0.2/Timer.java:506)

"vert.x-eventloop-thread-0" #19 prio=5 os_prio=0 cpu=127,32ms elapsed=76,90s tid=0x00007fde06895000 nid=0x157e runnable  [0x00007fdd2436d000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPoll.wait(java.base@13.0.2/Native Method)
	at sun.nio.ch.EPollSelectorImpl.doSelect(java.base@13.0.2/EPollSelectorImpl.java:120)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(java.base@13.0.2/SelectorImpl.java:124)
	- locked <0x000000070ea3d860> (a io.netty.channel.nio.SelectedSelectionKeySet)
	- locked <0x000000070ea3d808> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(java.base@13.0.2/SelectorImpl.java:136)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:62)
	at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:824)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
	at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1044)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(java.base@13.0.2/Thread.java:830)

"vert.x-eventloop-thread-1" #20 prio=5 os_prio=0 cpu=7,91ms elapsed=76,78s tid=0x00007fdd1009b000 nid=0x157f runnable  [0x00007fdd1dffa000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPoll.wait(java.base@13.0.2/Native Method)
	at sun.nio.ch.EPollSelectorImpl.doSelect(java.base@13.0.2/EPollSelectorImpl.java:120)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(java.base@13.0.2/SelectorImpl.java:124)
	- locked <0x000000070ea3df60> (a io.netty.channel.nio.SelectedSelectionKeySet)
	- locked <0x000000070ea3df08> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(java.base@13.0.2/SelectorImpl.java:136)
	at io.netty.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:62)
	at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:824)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
	at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1044)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(java.base@13.0.2/Thread.java:830)

"ADSI prefetch thread" #22 daemon prio=5 os_prio=0 cpu=245,68ms elapsed=76,12s tid=0x00007fde06b68800 nid=0x1580 waiting on condition  [0x00007fdd1dcf9000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@13.0.2/Native Method)
	- parking to wait for  <0x000000070e49c430> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@13.0.2/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@13.0.2/AbstractQueuedSynchronizer.java:2081)
	at java.util.concurrent.LinkedBlockingQueue.put(java.base@13.0.2/LinkedBlockingQueue.java:341)
	at org.nd4j.linalg.dataset.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:427)

"DeallocatorServiceThread_0" #23 daemon prio=5 os_prio=0 cpu=0,36ms elapsed=76,11s tid=0x00007fdd08005000 nid=0x1581 in Object.wait()  [0x00007fdd1dbf8000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@13.0.2/Native Method)
	- waiting on <0x000000070e4accb0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:155)
	- locked <0x000000070e4accb0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:176)
	at org.nd4j.linalg.memory.deallocation.DeallocatorService$DeallocatorServiceThread.run(DeallocatorService.java:123)

"DeallocatorServiceThread_1" #24 daemon prio=5 os_prio=0 cpu=0,21ms elapsed=76,11s tid=0x00007fdd08006800 nid=0x1582 in Object.wait()  [0x00007fdd1daf7000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@13.0.2/Native Method)
	- waiting on <0x000000070e4bd5c0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:155)
	- locked <0x000000070e4bd5c0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@13.0.2/ReferenceQueue.java:176)
	at org.nd4j.linalg.memory.deallocation.DeallocatorService$DeallocatorServiceThread.run(DeallocatorService.java:123)

"Attach Listener" #26 daemon prio=9 os_prio=0 cpu=2,95ms elapsed=0,10s tid=0x00007fddb4001000 nid=0x20f6 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"VM Thread" os_prio=0 cpu=56,70ms elapsed=80,18s tid=0x00007fde0426b000 nid=0x153a runnable  

"GC Thread#0" os_prio=0 cpu=198,52ms elapsed=80,20s tid=0x00007fde0408e000 nid=0x1535 runnable  

"GC Thread#1" os_prio=0 cpu=193,38ms elapsed=79,69s tid=0x00007fddc8001000 nid=0x154e runnable  

"GC Thread#2" os_prio=0 cpu=171,20ms elapsed=79,69s tid=0x00007fddc8003000 nid=0x154f runnable  

"GC Thread#3" os_prio=0 cpu=179,41ms elapsed=79,69s tid=0x00007fddc8004800 nid=0x1550 runnable  

"GC Thread#4" os_prio=0 cpu=138,73ms elapsed=79,69s tid=0x00007fddc8006000 nid=0x1551 runnable  

"GC Thread#5" os_prio=0 cpu=155,73ms elapsed=78,05s tid=0x00007fddc8038000 nid=0x156b runnable  

"GC Thread#6" os_prio=0 cpu=127,28ms elapsed=78,05s tid=0x00007fddc8039800 nid=0x156c runnable  

"G1 Main Marker" os_prio=0 cpu=3,65ms elapsed=80,20s tid=0x00007fde040af000 nid=0x1536 runnable  

"G1 Conc#0" os_prio=0 cpu=57,26ms elapsed=80,20s tid=0x00007fde040b0800 nid=0x1537 runnable  

"G1 Conc#1" os_prio=0 cpu=58,12ms elapsed=77,87s tid=0x00007fddd4001000 nid=0x156d runnable  

"G1 Conc#2" os_prio=0 cpu=58,71ms elapsed=77,87s tid=0x00007fddd4002800 nid=0x156e runnable  

"G1 Refine#0" os_prio=0 cpu=0,14ms elapsed=80,20s tid=0x00007fde041b7000 nid=0x1538 runnable  

"G1 Young RemSet Sampling" os_prio=0 cpu=14,97ms elapsed=80,20s tid=0x00007fde041b8800 nid=0x1539 runnable  
"VM Periodic Task Thread" os_prio=0 cpu=30,31ms elapsed=79,60s tid=0x00007fde056d5800 nid=0x1553 waiting on condition  

JNI global refs: 49, weak refs: 4354
```

**Comment by AlexDBlack** [comment]
cc @raver119 - any thoughts?
Looks like it might be a c++ deadlock on `Nd4jCpu.execSummaryStatsScalar`, but I can't see how/why that would happen (never seen this on any of our machines, or CI machines).
Java version and linux variant shouldn't make any difference AFAIK.

@vincenzocaselli what exactly is the processor model? I saw a couple of different possibilities for the Dell 5530

**Comment by raver119** [comment]
@vincenzocaselli so, you're using example as is, and it deadlocks for you?

**Comment by AlexDBlack** [comment]
@raver119 That's my understanding, yes, unmodified latest version of dl4j-examples (1.0.0-beta6)
I've tried that same example on my machines, and it's fine.
(We also run it as part of pre-release QA anyway)

**Comment by vincenzocaselli** [comment]
> Java version and linux variant shouldn't make any difference AFAIK.

As a side note I tried both with 
```
OpenJDK Runtime Environment (build 13.0.2+8)
OpenJDK 64-Bit Server VM (build 13.0.2+8, mixed mode)

```
and
```
Java(TM) SE Runtime Environment (build 13.0.1+9)
Java HotSpot(TM) 64-Bit Server VM (build 13.0.1+9, mixed mode, sharing)
```
with same results.


> @vincenzocaselli what exactly is the processor model? I saw a couple of different possibilities for the Dell 5530

Intel Core i7-8850H (Six Core 2.60GHz, 4.30GHz Turbo, 9MB 45W)


> @vincenzocaselli so, you're using example as is, and it deadlocks for you?

yes

Let me know if I can make other tests: different JVMs, change some env variable, ...

**Comment by raver119** [comment]
Can you please `export OMP_NUM_THREADS=1` and run again? Things will get slower, but we'll rule out parallelism.

**Comment by vincenzocaselli** [comment]
> Can you please `export OMP_NUM_THREADS=1` and run again? Things will get slower, but we'll rule out parallelism.

@raver119  done, but unfortunately the behavior is exactly the same

**Comment by raver119** [comment]
Hm. That doesn't sound realistic to me. Can you please disable UI stuff, and put regular score listener there with frequency of 1?

**Comment by vincenzocaselli** [comment]
@raver119 by skipping UI stuff it works like a charm
```
    public static void main(String[] args){
        //Get our network and training data
        MultiLayerNetwork net = UIExampleUtils.getMnistNetwork();
        DataSetIterator trainData = UIExampleUtils.getMnistData();

//        //Initialize the user interface backend
//        UIServer uiServer = UIServer.getInstance();
//
//        //Configure where the network information (gradients, activations, score vs. time etc) is to be stored
//        //Then add the StatsListener to collect this information from the network, as it trains
//        StatsStorage statsStorage = new FileStatsStorage(new File(System.getProperty("java.io.tmpdir"), "ui-stats.dl4j"));
//        int listenerFrequency = 1;
//        net.setListeners(new StatsListener(statsStorage, listenerFrequency));
//
//        //Attach the StatsStorage instance to the UI: this allows the contents of the StatsStorage to be visualized
//        uiServer.attach(statsStorage);

        net.setListeners(new ScoreIterationListener(1));
        
        //Start training:
        net.fit(trainData);

        //Finally: open your browser and go to http://localhost:9000/train
    }

```

```
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/vincenzo/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.25/8dacf9514f0c707cbbcdd6fd699e8940d42fb54e/slf4j-simple-1.7.25.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/vincenzo/.m2/repository/ch/qos/logback/logback-classic/1.1.7/logback-classic-1.1.7.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.SimpleLoggerFactory]
[main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for linear algebra: 1
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - *********************************** CPU Feature Check Warning ***********************************
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - Warning: Initializing ND4J with Generic x86 binary on a CPU with AVX/AVX2 support
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - Using ND4J with AVX/AVX2 will improve performance. See deeplearning4j.org/cpu for more details
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - Or set environment variable ND4J_IGNORE_AVX=true to suppress this warning
[main] WARN org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory - *************************************************************************************************
[main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for OpenMP BLAS: 1
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Linux]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [12]; Memory: [3.8GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
[main] INFO org.deeplearning4j.nn.multilayer.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 0 is 2.4596229357207373
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 1 is 2.3847859273214307
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 2 is 2.3786044546229355
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 3 is 2.2900389586334073
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 4 is 2.2641599509936485
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 5 is 2.248210364763382
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 6 is 2.055237813208146
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 7 is 2.020907858175258
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 8 is 1.9386540538316195
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 9 is 1.8051564138398883
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 10 is 1.6514443572336142
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 11 is 1.572192444812278
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 12 is 1.4436239291466388
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 13 is 1.0696110265790622
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 14 is 1.1613952072641363
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 15 is 0.8913948249016883
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 16 is 1.1365792501164045
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 17 is 0.9962026023624351
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 18 is 0.8859268674927635
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 19 is 0.8205417612492297
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 20 is 0.708769747733592
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 21 is 0.7118142633597823
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 22 is 0.6540580576891457
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 23 is 0.6822972891144343
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 24 is 0.7132086500773225
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 25 is 0.7072300808610156
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 26 is 0.6134240954977738
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 27 is 0.7122743342806406
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 28 is 0.7319481317854911
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 29 is 0.7265425042849846
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 30 is 0.5727326479706499
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 31 is 0.5091023160649354
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 32 is 0.49789919441507197
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 33 is 0.538915619645431
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 34 is 0.4144490042284773
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 35 is 0.444938702184474

```

**Comment by raver119** [comment]
Great. Now enable UI listener again, but keep score listener too.

**Comment by vincenzocaselli** [comment]
@raver119 now I see the training going on (sure enough it has always been working on background also before), but still the url is unreachable. Seems like the UI stuff doesn't start or something is blocking it on my system (firewall is off however).

**Comment by raver119** [comment]
So, there's no deadlock, and probably there never was any deadlock... Probably that's just operation that takes significant time and you just tend to see it more often there. We'll check it out.

**Comment by AlexDBlack** [comment]
@vincenzocaselli OK, can you try running this exact same example through maven or intellij instead of Eclipse IDE?
looking back over what you've posted this looks off
```
SLF4J: Found binding in [jar:file:/home/vincenzo/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.25/8dacf9514f0c707cbbcdd6fd699e8940d42fb54e/slf4j-simple-1.7.25.jar!/org/slf4j/impl/StaticLoggerBinder.class]
```

There shouldn't be anything gradle-related on the classpath in the examples. Suggests that maybe Eclipse IDE is causing problems with logging, and we're simply not seeing some other output (like an error saying port 9000 is already used, or something of that nature).
Vertx has definitely started according to the thread dump you posted earlier.

**Comment by vincenzocaselli** [comment]
@raver119 wow! A step forward: since I would have some problem using IntelliJ right now and don't know exactly how to use directly maven to run I followed your hint.
I therefore cleared my home .gradle/caches/ folder and now I see something different:
1. the console is messaging in black color (before was almost all in red), see below its content
2. the url is reachable!!! But it shows just "Resource not found"

```
o.n.l.f.Nd4jBackend - Loaded [CpuBackend] backend
o.n.n.NativeOpsHolder - Number of threads used for linear algebra: 1
o.n.l.c.n.CpuNDArrayFactory - *********************************** CPU Feature Check Warning ***********************************
o.n.l.c.n.CpuNDArrayFactory - Warning: Initializing ND4J with Generic x86 binary on a CPU with AVX/AVX2 support
o.n.l.c.n.CpuNDArrayFactory - Using ND4J with AVX/AVX2 will improve performance. See deeplearning4j.org/cpu for more details
o.n.l.c.n.CpuNDArrayFactory - Or set environment variable ND4J_IGNORE_AVX=true to suppress this warning
o.n.l.c.n.CpuNDArrayFactory - *************************************************************************************************
o.n.n.Nd4jBlas - Number of threads used for OpenMP BLAS: 1
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CPU]; OS: [Linux]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [12]; Memory: [3.8GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
o.d.n.m.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
o.d.u.VertxUIServer - Deeplearning4j UI server started at: http://localhost:9000
o.d.u.VertxUIServer - StatsStorage instance attached to UI: FileStatsStorage(/tmp/ui-stats.dl4j)
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 2.4596229357207373
o.d.o.l.ScoreIterationListener - Score at iteration 1 is 2.3847859273214307
o.d.o.l.ScoreIterationListener - Score at iteration 2 is 2.3786044546229355
o.d.o.l.ScoreIterationListener - Score at iteration 3 is 2.2900389586334073

```

**Comment by AlexDBlack** [comment]
@vincenzocaselli for maven execution, use this:
```
cd deeplearning4j-examples
mvn package
cd dl4j-examples
java -cp target/dl4j-examples-1.0.0-beta6-bin.jar org.deeplearning4j.examples.userInterface.UIExample
```

**Comment by vincenzocaselli** [comment]
@raver119 done with maven direct. Same result ... that is: url http://localhost:9000/train is unreachable.
But I noticed the message:
`o.d.u.VertxUIServer - Deeplearning4j UI server started at: http://localhost:9000`
so tried removing /train and ... it worked!
The final target Url is indeed http://localhost:9000/train/overview
but anyway just /train doesn't work.
Thank you @raver119 !!! 
I tried again from Eclipse and it is working also there.
Now I am trying to apply that to another project where I am using DL4J and where the problem of hanging is still there. I'll try to compare the two project for the differences and let you know if I can see the DL4J UI also there.
Thank you!

**Comment by AlexDBlack** [comment]
OK, great. I'll close this. :+1:

Only thing I think we can do here is provide a simple redirect from `/train` to `/train/overview`. I thought we already had that, but apparently not.
I'll open a separate issue for that.

**Comment by vincenzocaselli** [comment]
Hi @AlexDBlack, @raver119 
I am adding a comment here since the problem at the root of this post is back again.
Let me know if I should better open another issue and I'll do; for now I'll explain here below.
The problem was solved for the DL4L **UIExample**, but it is still present for a brand new project.

In fact I started (with Eclipse IDE, but don't think this can make a difference) a brand **new Maven project**.
Then added the following dependencies
```
 		<dependency>
			<groupId>org.nd4j</groupId>
			<artifactId>nd4j-native-platform</artifactId>
			<version>1.0.0-beta6</version>
		</dependency>
		<dependency>
			<groupId>org.deeplearning4j</groupId>
			<artifactId>deeplearning4j-core</artifactId>
			<version>1.0.0-beta6</version>
		</dependency>
  		<dependency>
			<groupId>org.deeplearning4j</groupId>
			<artifactId>deeplearning4j-ui</artifactId>
			<version>1.0.0-beta6</version>
		</dependency>
```
Then I created a very simple class for traning and testing against MNIST (here is the complete source code http://www.rcp-vision.com/build-your-first-neural-network-with-eclipse-deeplearning4j).
The code worked successfully.

Then I introduced UI stuff, but the URLs http://localhost:9000 or http://localhost:9000/train/overview
are unreachable and the console shows no errors.

I made many tests after comparing this brand new project and the one in deeplearning4j-examples (UIExample, it's pom.xml and it's parent pom.xml).
In one of these attempt I added the following dependency
```
        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-classic</artifactId>
            <version>1.1.7</version>
        </dependency>
```
With this simple step the console showed much more messages.
I am attaching these messages below in the hope that the error showed may help you to understand why the UI stuff is not working.
Thank you very much


```
17:05:32.420 [main] DEBUG io.netty.util.internal.PlatformDependent0 - -Dio.netty.noUnsafe: false
17:05:32.421 [main] DEBUG io.netty.util.internal.PlatformDependent0 - Java version: 13
17:05:32.421 [main] DEBUG io.netty.util.internal.PlatformDependent0 - sun.misc.Unsafe.theUnsafe: available
17:05:32.421 [main] DEBUG io.netty.util.internal.PlatformDependent0 - sun.misc.Unsafe.copyMemory: available
17:05:32.422 [main] DEBUG io.netty.util.internal.PlatformDependent0 - java.nio.Buffer.address: available
17:05:32.424 [main] DEBUG io.netty.util.internal.PlatformDependent0 - direct buffer constructor: unavailable
java.lang.UnsupportedOperationException: Reflective setAccessible(true) disabled
	at io.netty.util.internal.ReflectionUtil.trySetAccessible(ReflectionUtil.java:31)
	at io.netty.util.internal.PlatformDependent0$4.run(PlatformDependent0.java:224)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:312)
	at io.netty.util.internal.PlatformDependent0.<clinit>(PlatformDependent0.java:218)
	at io.netty.util.internal.PlatformDependent.isAndroid(PlatformDependent.java:272)
	at io.netty.util.internal.PlatformDependent.<clinit>(PlatformDependent.java:92)
	at io.netty.channel.nio.NioEventLoop.newTaskQueue0(NioEventLoop.java:285)
	at io.netty.channel.nio.NioEventLoop.newTaskQueue(NioEventLoop.java:156)
	at io.netty.channel.nio.NioEventLoop.<init>(NioEventLoop.java:138)
	at io.netty.channel.nio.NioEventLoopGroup.newChild(NioEventLoopGroup.java:138)
	at io.netty.channel.nio.NioEventLoopGroup.newChild(NioEventLoopGroup.java:37)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:84)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:58)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:47)
	at io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:59)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:78)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:73)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:60)
	at io.vertx.core.net.impl.transport.Transport.eventLoopGroup(Transport.java:148)
	at io.vertx.core.impl.VertxImpl.<init>(VertxImpl.java:143)
	at io.vertx.core.impl.VertxImpl.vertx(VertxImpl.java:92)
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:40)
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:32)
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:27)
	at io.vertx.core.Vertx.vertx(Vertx.java:75)
	at org.deeplearning4j.ui.VertxUIServer.getInstance(VertxUIServer.java:87)
	at org.deeplearning4j.ui.api.UIServer.getInstance(UIServer.java:58)
	at org.deeplearning4j.ui.api.UIServer.getInstance(UIServer.java:42)
	at it.rcpvision.dl4j.workbench.Step1MnistMLP.main(Step1MnistMLP.java:82)
17:05:32.424 [main] DEBUG io.netty.util.internal.PlatformDependent0 - java.nio.Bits.unaligned: available, true
17:05:32.425 [main] DEBUG io.netty.util.internal.PlatformDependent0 - jdk.internal.misc.Unsafe.allocateUninitializedArray(int): unavailable
java.lang.IllegalAccessException: class io.netty.util.internal.PlatformDependent0$6 cannot access class jdk.internal.misc.Unsafe (in module java.base) because module java.base does not export jdk.internal.misc to unnamed module @42a9a63e
	at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:376)
	at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:642)
	at java.base/java.lang.reflect.Method.invoke(Method.java:559)
	at io.netty.util.internal.PlatformDependent0$6.run(PlatformDependent0.java:334)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:312)
	at io.netty.util.internal.PlatformDependent0.<clinit>(PlatformDependent0.java:325)
	at io.netty.util.internal.PlatformDependent.isAndroid(PlatformDependent.java:272)
	at io.netty.util.internal.PlatformDependent.<clinit>(PlatformDependent.java:92)
	at io.netty.channel.nio.NioEventLoop.newTaskQueue0(NioEventLoop.java:285)
	at io.netty.channel.nio.NioEventLoop.newTaskQueue(NioEventLoop.java:156)
	at io.netty.channel.nio.NioEventLoop.<init>(NioEventLoop.java:138)
	at io.netty.channel.nio.NioEventLoopGroup.newChild(NioEventLoopGroup.java:138)
	at io.netty.channel.nio.NioEventLoopGroup.newChild(NioEventLoopGroup.java:37)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:84)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:58)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:47)
	at io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:59)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:78)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:73)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:60)
	at io.vertx.core.net.impl.transport.Transport.eventLoopGroup(Transport.java:148)
	at io.vertx.core.impl.VertxImpl.<init>(VertxImpl.java:143)
	at io.vertx.core.impl.VertxImpl.vertx(VertxImpl.java:92)
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:40)
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:32)
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:27)
	at io.vertx.core.Vertx.vertx(Vertx.java:75)
	at org.deeplearning4j.ui.VertxUIServer.getInstance(VertxUIServer.java:87)
	at org.deeplearning4j.ui.api.UIServer.getInstance(UIServer.java:58)
	at org.deeplearning4j.ui.api.UIServer.getInstance(UIServer.java:42)
	at it.rcpvision.dl4j.workbench.Step1MnistMLP.main(Step1MnistMLP.java:82)
17:05:32.425 [main] DEBUG io.netty.util.internal.PlatformDependent0 - java.nio.DirectByteBuffer.<init>(long, int): unavailable
17:05:32.426 [main] DEBUG io.netty.util.internal.PlatformDependent - sun.misc.Unsafe: available
17:05:32.438 [main] DEBUG io.netty.util.internal.PlatformDependent - maxDirectMemory: 4116709376 bytes (maybe)
17:05:32.438 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.tmpdir: /tmp (java.io.tmpdir)
17:05:32.438 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.bitMode: 64 (sun.arch.data.model)
17:05:32.439 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.maxDirectMemory: -1 bytes
17:05:32.439 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.uninitializedArrayAllocationThreshold: -1
17:05:32.439 [main] DEBUG io.netty.util.internal.CleanerJava9 - java.nio.ByteBuffer.cleaner(): available
17:05:32.440 [main] DEBUG io.netty.util.internal.PlatformDependent - -Dio.netty.noPreferDirect: false
17:05:32.446 [main] DEBUG io.netty.util.internal.PlatformDependent - org.jctools-core.MpscChunkedArrayQueue: available
17:05:32.476 [main] DEBUG io.netty.resolver.dns.DefaultDnsServerAddressStreamProvider - Default DNS servers: [/192.168.43.1:53] (sun.net.dns.ResolverConfiguration)
17:05:32.480 [main] DEBUG io.netty.util.NetUtil - -Djava.net.preferIPv4Stack: false
17:05:32.481 [main] DEBUG io.netty.util.NetUtil - -Djava.net.preferIPv6Addresses: false
17:05:32.482 [main] DEBUG io.netty.util.NetUtil - Loopback interface: lo (lo, 0:0:0:0:0:0:0:1%lo)
17:05:32.483 [main] DEBUG io.netty.util.NetUtil - /proc/sys/net/core/somaxconn: 128
17:05:33.096 [main] INFO org.deeplearning4j.ui.VertxUIServer - StatsStorage instance attached to UI: FileStatsStorage(/tmp/ui-stats.dl4j)
```

**Comment by AlexDBlack** [comment]
@vincenzocaselli so those stack traces you posted are likely irrelevant. You'll note they are debug level in netty, checking for certain JDK features - with properly configured logging (i.e., a logback.xml in the project resources directory like the link below) you wouldn't even see that.
https://github.com/eclipse/deeplearning4j-examples/blob/master/standalone-sample-project/src/main/resources/logback.xml


OK, three things we'll need here
(a) The _exact_ code you're using (with your network, listener configuration, UI-related code etc)
(b) The full output when running that code - I assume there's more than you posted? - in a gist - gist.github.com
(c) Confirmation of exactly how you are building and launching your project (via eclipse IDE? via maven + command line?)

I'm a little skeptical that it's something on our end, given we use the UI pretty frequently ourselves and haven't had any other reports of issues with it. But I should be able to help debug this once I have those details

**Comment by vincenzocaselli** [comment]
Hi @AlexDBlack ,
here they are:

(a) Java class & pom.xml
[Step1MnistMLP.java.txt](https://github.com/eclipse/deeplearning4j/files/4212409/Step1MnistMLP.java.txt)
[pom.xml.txt](https://github.com/eclipse/deeplearning4j/files/4212416/pom.xml.txt)

(b)
https://gist.github.com/vincenzocaselli/7fe545ed44551101228142822b7d95a8

(c) 
o.s. = Linux Arch 64bits (Manjaro Deepin 18.0.2)
Java JDK 13 (tried both Oracle and OpenJDK with same results)
Eclipse 2019-12 (Note: I use this system for daily work with Java+Eclipse for about a year now without any problem)
Then in short (here you can find more details and images http://www.rcp-vision.com/build-your-first-neural-network-with-eclipse-deeplearning4j):
* File, New, Project, Maven Project
* click to skip archetype selection, give a name for a Group Id and an Artifact Id
* open pom.xml and add block dependencies as in attachment
* create Java class as in attachment
* select Java class, right click, Run as Java Application
 
Please also note that without UI stuff the execution time is about  80 seconds, while here is 787
Let me know if you need other details.
Thank you

**Comment by vincenzocaselli** [comment]
Hi @AlexDBlack ,
I managed to follow **exactly the same steps above except** for just the following variants:
1) **IntelliJ** instead of Eclipse (but still on Linux Manjaro)
2) on **Windows10** instead of Linux Manjaro (but still with Eclipse)

and .. well, the results are exactly the same as above in both cases.

So there must be something wrong with the **pom.xml**.
What do you think?
Thank you 
Vincenzo

**Comment by treo** [comment]
@vincenzocaselli 
I've tried to reproduce your problem, but it works on all machines I've tried.

Can you step through `VertxUIServer.start` and see if something hinders it getting to the final log line?

Edit;
While trying to solve a similar issue with another user, we ran into a possible solution. Can you try adding this to your pom.xml?
```xml
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.freemarker</groupId>
                <artifactId>freemarker</artifactId>
                <version>2.3.29</version>
            </dependency>
        </dependencies>
    </dependencyManagement>
```

**Comment by vincenzocaselli** [comment]
Hi @treo !
wow, just by adding **freemarker** dependency make it work instantly!!!
Thank you, but what was happening?

**Comment by treo** [comment]
The problem is that the UI wants a newer version of freemarker, but something makes it so that an older version is included instead.
You're basically overwriting the freemarker version of the dependency that is already added. 

It is weird that the error message for it gets silently discarded.

**Comment by BrdEase** [comment]
Hi guys, I'm having the same issue as @vincenzocaselli 
I downloaded the full repo from https://github.com/eclipse/deeplearning4j-examples.git then ran it without any IDE for troubleshooting. Using maven on the command line as @AlexDBlack said, and the result is a little bit weird.
Despite the net running and even end training after some time (no errors or anything) the webhost at localhost:9000 never opens, i tried all the links you guys suggested, with no success.

Following what @treo said, I noticed that freemarker now came by default, anyway I tried 2.3.30 version as well with no luck.

I tried in windows 10 x64 on java 8 and 14.
And on Linux x64 on java 11.

When running mvn package some warnings show on the terminal:

```
[WARNING]   - META-INF.versions.9.module-info
[WARNING] maven-shade-plugin has detected that some class files are
 present in two or more JARs. When this happens, only one
 single version of the class is copied to the uber jar.
 Usually this is not harmful and you can skip these warnings, otherwise try to manually exclude artifacts based on
mvn dependency:tree -Ddetail=true and the above output.
 See http://maven.apache.org/plugins/maven-shade-plugin/
```

Running it with `java -cp dl4j-examples-1.0.0-beta7-shaded.jar org.deeplearning4j.examples.quickstart.features.userinterface.BasicUIExample` I get:

```
o.n.l.f.Nd4jBackend - Loaded [CpuBackend] backend
o.n.n.NativeOpsHolder - Number of threads used for linear algebra: 6
o.n.l.c.n.CpuNDArrayFactory - *********************************** CPU Feature Check Warning ***********************************
o.n.l.c.n.CpuNDArrayFactory - Warning: Initializing ND4J with Generic x86 binary on a CPU with AVX/AVX2 support
o.n.l.c.n.CpuNDArrayFactory - Using ND4J with AVX/AVX2 will improve performance. See deeplearning4j.org/cpu for more details
o.n.l.c.n.CpuNDArrayFactory - Or set environment variable ND4J_IGNORE_AVX=true to suppress this warning
o.n.l.c.n.CpuNDArrayFactory - *************************************************************************************************
o.n.n.Nd4jBlas - Number of threads used for OpenMP BLAS: 6
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CPU]; OS: [Windows 10]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [6]; Memory: [2,0GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
o.d.n.m.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]
o.d.u.VertxUIServer - Deeplearning4j UI server started at: http://localhost:9000
o.d.u.VertxUIServer - StatsStorage instance attached to UI: FileStatsStorage(C:\Users\Murilo\AppData\Local\Temp\ui-stats.dl4j)
o.d.u.VertxUIServer - StatsStorage instance attached to UI: InMemoryStatsStorage(uid=a304c5ff)
```

(I dont know why the line endings are not working so I put it on pastebin as well: [link](https://pastebin.pl/view/34be76ce)

and then after some time it prints this:
```
o.d.u.VertxUIServer - Deeplearning4j UI server is auto-stopping after thread (name: main) died.
o.d.u.VertxUIServer - Deeplearning4j UI server stopped.
```

This warning and output is consistent over the OS's I tested.

I really don't know what is happening, and I want to help to sort this out. Any additional information will be sent as soon as possible.

**Comment by treo** [comment]
I've fixed the line endings for you. You've got to use tripple backticks to get it to work like that instead of the single backticks you've used.

What you are seeing is intended behavior in the current version. When the training is finished, it also stops the UI.

If you want to keep it running, you just have to keep the main thread alive. For example you could read from stdin and wait for a specific key to be entered or something like that, the easiest way would be to just let the thread sleep if you just want to look around for a few minutes.

**Comment by BrdEase** [comment]
Thanks for the line endings, I didn't know that.
And also by quick answer, appreciate it.

My problem is not that it stops running, I'm saying that I can't access the UI in the browser as expected. I just posted the full output to document it all.

Actually I have tried to add a `Thread.sleep(999999)` in the end, and it won't let the thread die, but I still cant access the web UI.

Sorry for not specifying that before. 
Is there any other thing to try?

**Comment by treo** [comment]
Well, that led me on a wild goose chase... Apparently vertx swallows exceptions when it is configured to support HTTP2 upgrades....

That weird behavior (Port is open, but it just hangs when connecting) is yet again a dependency version clash. 

You'll need to add the following to the `dependencyManagement` section:
```
            <dependency>
                <groupId>io.netty</groupId>
                <artifactId>netty-buffer</artifactId>
                <version>4.1.48.Final</version>
            </dependency>
```

**Comment by BrdEase** [comment]
I can't even describe how awesome you are. That's a Christmas gift. I fought so hard for so long with this but I didn't have the knowledge to solve it. Thank you.

**Comment by krisdesalpes** [comment]
Hi,

i have the same problem with version beta7 but when i look in the pom.xml i see @treo suggestions to add depency but it did not work.

i see now this issue closed, but i got the problem. I am lost...

thanks by advance

**Comment by agibsonccc** [comment]
@krisdesalpes  mind opening a new issue with  a bit more context? Thanks!

**Comment by vkazlouski** [comment]
Just wanted to mention that in order to fix this "/train url is unreachable" issue in the latest codebase you need to have both netty-common and netty-buffer dependencies set to 4.1.48.Final (As per @treo 's comment).  So your dependencyManagement tag will look in the following way:
```
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.freemarker</groupId>
                <artifactId>freemarker</artifactId>
                <version>2.3.29</version>
            </dependency>
	    <dependency>
                <groupId>io.netty</groupId>
                <artifactId>netty-common</artifactId>
                <version>4.1.48.Final</version>
            </dependency>            
	     <dependency>
                <groupId>io.netty</groupId>
                <artifactId>netty-buffer</artifactId>
                <version>4.1.48.Final</version>
            </dependency>
        </dependencies>
    </dependencyManagement>
```
Env: Eclipse Version: 2019-12 (4.14.0) / OS: Linux, v.5.4.0-58-generic, x86_64 / mvn: 3.6.3 / java version: 15.0.1

In the current version of dl4j-examples/pom.xml "netty-buffer" is missing. This is where it comes from:
datavec-local->(datavec-arrow->arrow-vector->netty-buffer:4.1.22 | datavec-arrow->arrow-memory->netty-buffer:4.1.22)
